File: lc/c/002.c
#define EMPTY 0x80808080

int lengthOfLongestSubstring(char * s){
    int ret = 0;
    int start = 0, slen = strlen(s);
    int prev[128];
    memset(prev, 0x80, sizeof prev);
    for (int i = 0; i < slen; ++i) {
        if (prev[s[i]] != EMPTY && prev[s[i]] >= start)
            start = prev[s[i]] + 1;
        prev[s[i]] = i;
        ret = i-start+1 > ret ? i-start+1 : ret;
    }
    return ret;
}

File: lc/py/002.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ret = 0
        start = 0
        prev = {}
        for i, c in enumerate(s):
            if c in prev and prev[c] >= start:
                start = prev[c]+1
            prev[c] = i
            ret = max(ret, i-start+1)
        return ret
